includes/plugins/class-woocommerce-memberships.php

```diff
@@ -207,6 +235,14 @@ public static function add_user_to_list( $plan, $args ) {
 			}
 		}
 
+		/**
+		 * In case the user was previously removed from the lists, we want to resubscribe them only to the lists they were in.
+		 */
+		$pre_existing_lists = self::get_user_lists_on_deactivation( $user_id, $user_membership->get_id() );
+		if ( ! empty( $pre_existing_lists ) ) {
+			$lists_to_add = array_intersect( $lists_to_add, $pre_existing_lists );
```

Comment:
Non-blocking suggestion, as there's no need to keep that after the subscription has been restored:

```suggestion
			$lists_to_add = array_intersect( $lists_to_add, $pre_existing_lists );
			delete_user_meta( $user_id, self::SUBSCRIBED_ON_DEACTIVATION_META_KEY );
```

Any time this happens again the meta should have the most recent subscription information, so it's also safer to not keep it in the database.

Reply:
If a membership switch status from a status considered active to another status considered active, we would end up subscribing them to all lists. For example, if you switch from `free_trial` to `active`. Or even `pending` is considered active. See `WC_Memberships_User_Memberships::get_active_access_membership_statuses()`

Reply:
But we need to make sure to update it when the membership is set to inactive. Looking at the code now, it seems tere could be some situations where we would keep outdated values there, let me have a closer look

Reply:
So if they unsubscribe and switch to another "active" status, they'll get re-subscribed to the list they've previously opted out?

Reply:
No... that's what we are trying to avoid

Reply:
> Looking at the code now, it seems tere could be some situations where we would keep outdated values there, let me have a closer look

This is what I meant. If they unsubscribe and the subscription status changes from "active" to "active", they'll get resubscribed.

Reply:
what I'm saying is I think we need to remove this check _> https://github.com/Automattic/newspack-newsletters/pull/1377/files#diff-257035a86886b24eaffe64a9bbaf2b138a59b755e71ebf379317a588e7a28b5fR170

And store an empty list in case they are not subscribed to any list.

This is to cover the case in which they:
* are granted a membership
* the membership is set inactive (stores the current lists -> say they have one)
* the membership is set to active again -> resubscribe to one list
* They unsubscribe from that one list
* the membership is set o inactive again (with the current check, we would keep the old value there, with one list)
* the membership is set to active again (if the old meta is there, they would get subscribed to that one list again)

Reply:
> if the old meta is there, they would get subscribed to that one list again

this is what I'm trying to say 😅

Reply:
https://github.com/Automattic/newspack-newsletters/pull/1377/commits/bae7c857e22a63e1face44a797a0e7f607000736

It's basically what you were suggesting, but doing that when the membership is set to inactive. Always update the info.

Actually. Also need this -> https://github.com/Automattic/newspack-newsletters/pull/1377/commits/e149a74b2065715b6149790188c4d6df90e27770

Reply:
Are there no cases in which the membership goes from "active" to "active"? If the meta is only updated when they are removed from a list, the "active" to "active" switch would also trigger a potentially outdated value.

Reply:
hmm you're right. Thinking...

Reply:
Can't the method bail if the previous status is "active"?

Reply:
We don't have the previous status in that context. And it took me a long while to figure out what hooks to use, I didn't want to explore changing it to another hook

Reply:
An additional user meta can store the last known membership status so you can check against that.

Reply:
status transitions run before `save_post` which is where the other hook is triggered. This seems to work on my tests:

https://github.com/Automattic/newspack-newsletters/pull/1377/commits/3b3d96c2a1c5dff44ecb25fd705eda1891b23a01

Additional testing instructions

* add user to membership
* as the user, unsubscribe from a list
* change the membership from an active state to another active state
* see that nothing is done

Reply:
I have not tested it yet, but how does the `$previous_statuses` is populated when switching from "active" to "active" if it's happening inside the `remove_user_from_list()` method, which is not called in that context?

Reply:
Nevermind, it is called but bails [here](https://github.com/Automattic/newspack-newsletters/blob/3b3d96c2a1c5dff44ecb25fd705eda1891b23a01/includes/plugins/class-woocommerce-memberships.php#L142-L144)

Reply:
Yes, maybe that method deserved a different name now... ?